server/src/main/java/org/apache/iotdb/db/query/fill/PreviousFill.java

```diff
@@ -59,28 +84,184 @@ public long getBeforeRange() {
+      if (endtimeContainedByTimeFilter(timeseriesMetadata.getStatistics())) {
         break;
       }
     }
+  }
+
+  private TimeValuePair getChunkLastPoint(ChunkMetadata chunkMetaData) throws IOException {
+    TimeValuePair lastPoint = new TimeValuePair(0, null);
+    Statistics chunkStatistics = chunkMetaData.getStatistics();
+
+    if (endtimeContainedByTimeFilter(chunkStatistics)) {
+      return constructLastPair(
+          chunkStatistics.getEndTime(), chunkStatistics.getLastValue(), dataType);
+    }
+    List<IPageReader> pageReaders = FileLoaderUtils.loadPageReaderList(chunkMetaData, timeFilter);
+    for (int i = pageReaders.size() - 1; i >= 0; i--) {
+      IPageReader pageReader = pageReaders.get(i);
+      Statistics pageStatistics = pageReader.getStatistics();
+      if (!timeFilter.satisfy(pageStatistics)) {
+        continue;
+      }
+      if (endtimeContainedByTimeFilter(pageStatistics)) {
+        lastPoint = constructLastPair(
+            pageStatistics.getEndTime(), pageStatistics.getLastValue(), dataType);
+      } else {
+        BatchData batchData = pageReader.getAllSatisfiedPageData();
+        lastPoint = batchData.getLastPairBeforeOrEqualTimestamp(queryTime);
+      }
+      break;
+    }
+    return lastPoint;
+  }
+
+  private boolean shouldUpdate(long time, long version, long newTime, long newVersion) {
+    return time < newTime || (time == newTime && version < newVersion);
+  }
+
+  private PriorityQueue<TsFileResource> sortUnSeqFileResourcesInDecendingOrder(
+      List<TsFileResource> tsFileResources) {
+    PriorityQueue<TsFileResource> unseqTsFilesSet =
+        new PriorityQueue<>(
+            (o1, o2) -> {
+              Map<String, Long> startTimeMap = o1.getEndTimeMap();
+              Long minTimeOfO1 = startTimeMap.get(seriesPath.getDevice());
+              Map<String, Long> startTimeMap2 = o2.getEndTimeMap();
+              Long minTimeOfO2 = startTimeMap2.get(seriesPath.getDevice());
+
+              return Long.compare(minTimeOfO2, minTimeOfO1);
+            });
+    unseqTsFilesSet.addAll(tsFileResources);
+    return unseqTsFilesSet;
+  }
 
-    if (beforePair != null) {
-      beforePair.setTimestamp(queryTime);
-    } else {
-      beforePair = new TimeValuePair(queryTime, null);
+  private PriorityQueue<ChunkMetadata> sortUnseqChunkMetadatasByEndtime() throws IOException {
+    PriorityQueue<ChunkMetadata> chunkMetadataList =
+        new PriorityQueue<>(
+            (o1, o2) -> {
+              long endTime1 = o1.getEndTime();
+              long endTime2 = o2.getEndTime();
+              if (endTime1 < endTime2) {
+                return 1;
+              } else if (endTime1 > endTime2) {
+                return -1;
+              }
+              return Long.compare(o2.getVersion(), o1.getVersion());
+            });
+    while (!unseqTimeseriesMetadataList.isEmpty()) {
+      chunkMetadataList.addAll(unseqTimeseriesMetadataList.remove(0).loadChunkMetadataList());
```

Comment:
no need to remove(0),just traverse is ok. Otherwise, use LinkedList for higher efficiency